sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -986,11 +996,13 @@ public boolean advance() throws IOException {
               rawRecord.topic(),
               rawRecord.partition(),
               rawRecord.offset(),
+              consumerSpEL.getRecordTimestamp(rawRecord),
               decode(rawRecord.key(), source.spec.getKeyCoder()),
               decode(rawRecord.value(), source.spec.getValueCoder()));
 
           curTimestamp = (source.spec.getTimestampFn() == null)
-              ? Instant.now() : source.spec.getTimestampFn().apply(record);
+              ? new Instant(record.getTimestamp())
```

Comment:
So joining a bit late to this party.. I might ask the same questions, but I will anyway, in code-context:
A pre-defined WM function takes precedent, otherwise you use the consumer timestamp which might be Kafka-native (0.10) or simply current time ? 
Is this desired ? to let the user/IO override Kafka's timestamp ?